src/com/ms/silverking/cloud/dht/benchmark/ycsb/GroupingSilverkingDB.java

```diff
@@ -234,199 +235,199 @@ private void read(List<ReadOp> readOps) {
         }
     }
     */
-    
-    private static final LightLinkedBlockingQueue<ReadOp>  readOps; 
 
-    public int read(String table, String key, Set<String> fields, HashMap<String, ByteIterator> values) {
+  private static final LightLinkedBlockingQueue<ReadOp> readOps;
+
+  public int read(String table, String key, Set<String> fields, HashMap<String, ByteIterator> values) {
+    try {
+      ReadOp readOp;
+      boolean complete;
+
+      readOp = new ReadOp(table, key, fields, values);
+      readOps.put(readOp);
+      if (debug) {
+        System.out.println("waiting for: " + key + "\t" + readOp);
+      }
+      complete = readOp.waitForCompletion(TimeUnit.SECONDS, 5);
+      if (debug) {
+        System.out.println("back from wait: " + key + " " + complete + "\t" + readOp);
+      }
+      return readOp.returnCode;
+    } catch (InterruptedException ie) {
+      throw new RuntimeException("Unsupported interruption");
+    }
+  }
+
+  private static class Worker implements Runnable {
+    private final ReadOp[] groupOps;
+    private final SynchronousNamespacePerspective<String, Map> syncNSP;
+
+    private static final int maxGroupSize = 1024;
+
+    private Worker(SynchronousNamespacePerspective<String, Map> syncNSP) {
+      this.syncNSP = syncNSP;
+      groupOps = new ReadOp[maxGroupSize];
+    }
+
+    public void run() {
+      while (true) {
         try {
-            ReadOp  readOp;
-            boolean complete;
-            
-            readOp = new ReadOp(table, key, fields, values);
-            readOps.put(readOp);
-            if (debug) {
-                System.out.println("waiting for: "+ key +"\t"+ readOp);
-            }
-            complete = readOp.waitForCompletion(TimeUnit.SECONDS, 5);
-            if (debug) {
-                System.out.println("back from wait: "+ key +" "+ complete +"\t"+ readOp);
-            }
-            return readOp.returnCode;
+          int groupSize;
+
+          groupSize = readOps.takeMultiple(groupOps);
+          read(groupOps, groupSize);
         } catch (InterruptedException ie) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)